src/odemis/odemisd/main.py

```diff
@@ -328,71 +329,98 @@ def _terminate_all_alive(self):
         It also stops the containers, once no component is running in them.
         """
         mic = self._instantiator.microscope
-        alive = set(mic.alive.value)
-
-        # create a "graph" child->parents
-        parents = {c: set() for c in alive}  # comp -> set of comps (all its parents)
-        for comp in alive:
-            # TODO: if a component was created by delegation, use its creator
-            # instead of itself as "used" by the other components, to ensure
-            # they might not be terminated too early.
+
+        # create a "graph" dependency->parents
+        self._parents = {c: set() for c in set(mic.alive.value)}  # comp -> set of comps (all the comps that depend on it)
+        children_comps = []
+        for comp in self._parents:
             try:
-                for child in comp.children.value:
-                    # TODO: do not add the creator
-                    parents[child].add(comp)
+                for dep in comp.dependencies.value:
+                    # if a component was created by delegation, use its creator
+                    # instead of itself as "used" by the other components, to ensure
+                    # they might not be terminated too early.
+                    for c in self._parents:
```

Comment:
It should be possible to just use `dep.parent.value`, and if it's not None, it's the right parent.
(but double that it actually works, and maybe it's safer to not rely on this "back link" to work)